pilot/pkg/bootstrap/server.go

```diff
@@ -560,42 +555,110 @@ func (s *Server) initServiceControllers(args *PilotArgs) error {
 		}
 	}
 
-	s.serviceController = serviceControllers
+	s.ServiceController = serviceControllers
 
 	// Defer running of the service controllers.
 	s.addStartFunc(func(stop chan struct{}) error {
-		go s.serviceController.Run(stop)
+		go s.ServiceController.Run(stop)
 		return nil
 	})
 
 	return nil
 }
+func initMemoryRegistry(s *Server, serviceControllers *aggregate.Controller) {
+	discovery1 := mock.NewDiscovery(
+		map[string]*model.Service{
+			//			mock.HelloService.Hostname: mock.HelloService,
+		}, 2)
+
+	s.MemoryServiceDiscovery = discovery1
+
+	discovery2 := mock.NewDiscovery(
+		map[string]*model.Service{
+			//			mock.WorldService.Hostname: mock.WorldService,
+		}, 2)
+
+	registry1 := aggregate.Registry{
+		Name:             serviceregistry.ServiceRegistry("mockAdapter1"),
+		ServiceDiscovery: discovery1,
+		ServiceAccounts:  discovery1,
+		Controller:       &mockController{},
+	}
+
+	registry2 := aggregate.Registry{
+		Name:             serviceregistry.ServiceRegistry("mockAdapter2"),
+		ServiceDiscovery: discovery2,
+		ServiceAccounts:  discovery2,
+		Controller:       &mockController{},
+	}
+	serviceControllers.AddRegistry(registry1)
+	serviceControllers.AddRegistry(registry2)
+}
 
 func (s *Server) initDiscoveryService(args *PilotArgs) error {
 	environment := model.Environment{
 		Mesh:             s.mesh,
 		IstioConfigStore: model.MakeIstioStore(s.configController),
-		ServiceDiscovery: s.serviceController,
-		ServiceAccounts:  s.serviceController,
+		ServiceDiscovery: s.ServiceController,
+		ServiceAccounts:  s.ServiceController,
 		MixerSAN:         s.mixerSAN,
 	}
 
 	// Set up discovery service
 	discovery, err := envoy.NewDiscoveryService(
-		s.serviceController,
+		s.ServiceController,
 		s.configController,
 		environment,
 		args.DiscoveryOptions)
 	if err != nil {
 		return fmt.Errorf("failed to create discovery service: %v", err)
 	}
+	s.DiscoveryService = discovery
+
+	// For now we create the gRPC server sourcing data from Pilot's older data model.
+	s.initGrpcServer()
+	s.EnvoyXdsServer = envoyv2.NewDiscoveryServer(discovery, s.GRPCServer)
+
+	s.HTTPServer = &http.Server{
+		Addr:    ":" + strconv.Itoa(args.DiscoveryOptions.Port),
+		Handler: discovery.RestContainer}
+
+	addr := s.HTTPServer.Addr
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return err
```

Comment:
I couldn't find in the call chain where those errors get logged.
It would probably help troubleshooting to log something near/where the error occured?

Reply:
Code moved from discovery without changes :-)

Reply:
(DiscoveryService.Start)

Agree with the comment - but additional improvements can be made in separate PRs, not trying to fix all the problems in one.